### ISSUE TITLE
`@TinyJit` produces wrong gradients when a fresh BUFFER created by .contiguous() materializes a non-contiguous view, when memory planner is enabled.

### ISSUE THREAD

**Issue by Cyberistic** [tinygrad/tinygrad#16571]
Correct me if I'm missing something, still learning and already spent +12 hours on this issue (I will send the bill to geohot :p). This surfaced while working on DPSGD with ghost-clipping.

When `@TinyJit` does two consecutive `backward()` calls where the loss is a per-example tensor, the first JIT capture produces correct gradients but every subsequent replay produces wrong gradients.  This only happens when memory planner is enabled (NO_MEMORY_PLANNER=0). 

If the loss is already a scalar (`.mean()` then `.backward()`), or if `backward()` is called only once, the bug does not appear. The first replay is always wrong, every replay after that is the same wrong answer.

Specifically, these must be present (in my tests):

1. A **per-example** loss of shape `(B,)`, for example,  `gather`-based sparse CE
2. **Two consecutive `backward()` calls** inside the JIT (the second one can be a scalar multiple of the first, e.g. `(loss * 0.5).sum().backward()`)
3. The forward pass must allocate at least one **fresh BUFFER** via `.contiguous()` on a non-contiguous view. The simplest triggers are `X.permute(0, 3, 1, 2).contiguous()` (NHWC -> NCHW) or `X.transpose(2, 3).contiguous()` (any non-contig view -> fresh BUFFER). 

A plain `.contiguous()` on already-contiguous data is a no-op and does **not** trigger the bug. A `.permute()` without `.contiguous()` also does not trigger it, because no new BUFFER is materialized.

- tinygrad version: `0.13.0`
- Device: Tried both `CPU` and `Metal`. 
- Running on Mac M2

## Root cause

*write-after-read* issue in memory planner.  This issue does not reproduce when (NO_MEMORY_PLANNER=1). 

`memory_plan_rewrite` in `tinygrad/schedule/memory.py` reuses the memory of intermediate `BUFFER`s whose lifetimes it thinks do not overlap. The `held_bufs` set passed in by `TinyJit.__call__` (in `tinygrad/engine/jit.py`) only contains:

- pre-existing live buffers (`set(buffers)`)
- the `.buf_uop` of returned parameters (`get_parameters(ret)`)

It does **not** contain the BUFFERs that the JIT capture *itself* creates for the intermediate values of the new linear. So those BUFFERs go through the memory planner. The planner computes their lifetimes using the linear toposort and concludes that two non-overlapping BUFFERs can share one allocator slot.

But for a per-example loss + two `backward()`s, the second backward's `.grad` accumulation reads from BUFFERs that the planner has already freed. The planner reuses those slots for BUFFERs that are alive in the replay, so by the time the `.grad` reads them, they contain unrelated data.

I assume the first capture (`self.cnt == 1`) is correct because the planner runs on the *replay's* linear but the data layout being read is still the capture's. The replay is wrong because the capture's intermediate buffers are no longer populated when the planner re-runs.

We can see this clearly in the repro script:
```
linear has 15 BUFFERs total, 10 held (skipped), 5 planned (suballocated)
```

Seems like the 5 suballocated BUFFERs are the ones whose memory is being reused across the two backward() kernels.

## My fix

I'm not sure if this is the correct way to go around this, but here's what I did (other option was to disable memory planner). Hoping someone smarter than me will see what's up.

In `TinyJit.__call__` (around line 32 of `tinygrad/engine/jit.py`), expand `held_bufs` to include *every* `Ops.BUFFER` in the captured linear, not just pre-existing ones and parameter buffers:

```python
held_bufs = set(buffers) | {t.uop.buf_uop for t in get_parameters(ret) if t.uop.buf_uop.op is Ops.BUFFER}

# FIX: hold all BUFFERs in the captured linear so the memory planner
# does not suballocate intermediates that subsequent .grad reads depend on.
for n in big_linear.toposort():
    if n.op is Ops.BUFFER:
        held_bufs.add(n)

linear = jit_lower(big_linear, held_bufs, input_buf_uops)
```

## Repro

```python
# repro.py
import numpy as np
from tinygrad import Tensor, TinyJit, nn, Context


def per_example_sparse_ce(logits, y):
    lse = logits.logsumexp(axis=-1, keepdim=True)
    return -(logits - lse).gather(1, y.reshape(-1, 1)).reshape(-1)


def step(conv, X, y):
    """Forward + 2x backward."""

    # .contiguous() materializes the permuted view into a fresh BUFFER, without it, conv(Xn) reads the permuted view directly
    # so no new BUFFER is created for the planner to suballocate?
    Xn = X.permute(0, 3, 1, 2).contiguous() if X.shape[-1] == 3 else X

    logits = conv(Xn).relu().max_pool2d(2).flatten(1)
    loss = per_example_sparse_ce(logits, y)

    loss.sum().backward() # First backward: correct, populates .grad for the capture.
    conv.weight.grad = None
    (loss * 0.5).sum().backward() # Second backward: wrong, reads from corrupted buffers that the planner has suballocated.
    return loss.mean().realize()


B, C, H, W = 16, 10, 100, 100
np.random.seed(42); Tensor.manual_seed(42)
conv = nn.Conv2d(3, C, kernel_size=3, padding=1)
X = Tensor(np.random.randn(B, H, W, 3).astype(np.float32))
y = Tensor(np.random.randint(0, C, size=B).astype(np.int32))

with Context(DEV='CPU'):
    # Eager reference
    step(conv, X, y)
    ref = conv.weight.grad.numpy().copy()
    print(f"Reference: conv.weight.grad.sum() = {ref.sum():+.6f}")

    # JIT: same `step`, same `conv`, same inputs
    jit_step = TinyJit(Tensor.train()(step))

    for s in range(8):
        jit_step(conv, X, y)
        g = conv.weight.grad.numpy()
        ok = np.allclose(g, ref, rtol=1e-4, atol=1e-5)
        print(f"step {s}: conv.weight.grad.sum() = {g.sum():+.6f}  matches_ref = {ok}")
```

### (1) Planner ON, no fix (default)

Command: `python repro.py`

```
Reference: conv.weight.grad.sum() = +2.937127
step 0: conv.weight.grad.sum() = +2.937127  matches_ref = True
step 1: conv.weight.grad.sum() = +0.000000  matches_ref = False
step 2: conv.weight.grad.sum() = +0.000000  matches_ref = False
step 3: conv.weight.grad.sum() = +0.000000  matches_ref = False
step 4: conv.weight.grad.sum() = +0.000000  matches_ref = False
step 5: conv.weight.grad.sum() = +0.000000  matches_ref = False
step 6: conv.weight.grad.sum() = +0.000000  matches_ref = False
step 7: conv.weight.grad.sum() = +0.000000  matches_ref = False
```

### (2) Planner OFF (`NO_MEMORY_PLANNER=1`)

Command: `NO_MEMORY_PLANNER=1 python repro.py`

```
Reference: conv.weight.grad.sum() = +2.937127
step 0: conv.weight.grad.sum() = +2.937127  matches_ref = True
step 1: conv.weight.grad.sum() = +2.937127  matches_ref = True
step 2: conv.weight.grad.sum() = +2.937127  matches_ref = True
step 3: conv.weight.grad.sum() = +2.937127  matches_ref = True
step 4: conv.weight.grad.sum() = +2.937127  matches_ref = True
step 5: conv.weight.grad.sum() = +2.937127  matches_ref = True
step 6: conv.weight.grad.sum() = +2.937127  matches_ref = True
step 7: conv.weight.grad.sum() = +2.937127  matches_ref = True
```

### (3) Planner ON, with the fix I mentioned above

Command: `python repro.py` after applying the `held_bufs` expansion
shown above.

```
Reference: conv.weight.grad.sum() = +2.937127
step 0: conv.weight.grad.sum() = +2.937127  matches_ref = True
step 1: conv.weight.grad.sum() = +2.937127  matches_ref = True
step 2: conv.weight.grad.sum() = +2.937127  matches_ref = True
step 3: conv.weight.grad.sum() = +2.937127  matches_ref = True
step 4: conv.weight.grad.sum() = +2.937127  matches_ref = True
step 5: conv.weight.grad.sum() = +2.937127  matches_ref = True
step 6: conv.weight.grad.sum() = +2.937127  matches_ref = True
step 7: conv.weight.grad.sum() = +2.937127  matches_ref = True
```

Same as (2), but the planner still runs. It just finds nothing to suballocate within the JIT's captured linear, while non-JIT paths retain full suballocation.

**Comment by discobot** [comment]
Pinned down the mechanism: the memory planner rewrites the BUFFER read by the externally-held lazy `conv.weight.grad` graph (assigned during capture by `backward()`) into an arena slice, so replays never write the original BUFFER and realizing `.grad` reads freshly allocated memory — hence exactly 0.0. Fix in #16588: hold every BUFFER reachable from a live Tensor, so truly JIT-internal buffers still get planned.
